enhancements/machine-api/machine-api-usage-telemetry.md

```diff
@@ -0,0 +1,340 @@
+across several users or platforms provides opportunities to create alerts for
+users of potential outage events.
+
+**Number of Nodes Covered by the Machine Health Check**
+
+Epics: Engineering Insights, Proactive Support
+
+This metric will give us insight into user experiences operating the Machine Health
+Check. For example, users might think they are utilizing the Machine Health
+Check correctly only to find out that it has been misconfigured and not
+performing its role. If users have a large cluster with the health check
+enabled, but not covering any nodes, this might be an opportunity to reach out
+and see if we can help with their configuration. This can also help us identify
+areas where increased education will have value.
+
+**Failed Node Cleanups**
+
+Epics: Proactive Support
+
+This metric will help us identify conditions where the Cluster Autoscaler is
+silently cleaning up failied node creations. We should trigger an alert if the
+autoscaler cleans up more than X hosts in Y time period. This alert would be
+useful for users that might not realize they’re having these infrastructure problems.
+This is another case where signals across multiple clusters might indicate
+that a cloud or network area is having problems.
+
+For example, our internal continuous integration infrastructure team had a
+situation where two MachineSets had been configured for autoscaling.
+One MachineSet would scale up without issue, then the other would not. While
+this was happening, the Machines would fail, alarms would fire, and the autoscaler
+would dispose of the nodes that failed to come online. These failed nodes were
+being silently cleaned up by the autoscaler.
+
+**Total Scale Events**
+
+Epics: Demonstration of Value
+
+This metric will help us to show the level of activity that the Cluster
+Autoscaler is performing. Collecting data on the total number of scale up and
+scale down events, possibly as two individual metrics, we could correlate with
+instance pricing to demonstrate how much using the autoscaler saves our users.
+This can also serve as a good proxy for how many users are actually using the
+autoscaler in a meaningful way, not just who has turned it on.
+
+**Unable to Scale Events**
+
+Epics: Proactive Support
+
+This metric will help us identify situations when the Cluster Autoscaler is
+unable to perform scaling operations. To maximize the value of this data, we
+will want to report categories based on the different types of failures.
+Events related to min/max threshold on MachineSets probably don’t need
+alerting, but the inability to scale for other reasons (eg, hitting max CPU threshold or
+memory, but haven’t hit max node count) are useful to collect. These events can
+show opportunities to inform users about resource limitations in their clusters.
+
+**Machines with Old Delete Timestamp**
+
+Epics: Proactive Support
+
+This metric can show us stale resources that exist within clusters. By collecting
```

Comment:
The proposed alert feels like alerting on cause already. I would prefer alerting on the symptom of why we get stale resources so we prevent them from happening. See more about cause vs symptom based alerting here -> https://prometheus.io/docs/practices/alerting/

Reply:
that makes sense to me. this specific use case seems like a chicken/egg problem at some level. iirc, this metric story is about what we observe in the field (eg at some times clusters have these machines caught in stale deletion states) and adding some analysis to the failures that happens. we know that there is a root cause _somewhere_ but do not have an understanding of how it happened yet, but we know we have machine that get caught in this state. i _think_ that was what this metric was driving towards.

perhaps @michaelgugino could add some extra context here

Reply:
I disagree this would be alerting for 'cause' instead of 'symptom.'  The 'symptom' is a machine with a delete timestamp older than X.  The cause could be any number of reasons such as expired cloud credentials, undrainable node, machine-controller died, etc.

Reply:
https://prometheus.io/docs/practices/alerting/ links [this doc][1], which has a section on *Alerting from the spout* that describes cause vs. symptom applying to external users consuming cluster-provided services.  Stale machines seem far from the spout.  It make sense to me as a warn-level alert, just not a wake-me-at-midnight critical alert.  I don't see anything in this enhancement today that specifically calls out new alerts or their intended levels.

[1]: https://docs.google.com/a/boxever.com/document/d/199PqyG3UsyXlwieHaqbGiWVa8eMWi8zzAn0YfcApr8Q

Reply:
>  I don't see anything in this enhancement today that specifically calls out new alerts or their intended levels.

i'm still gathering the details for the alerts, hoping to add them today.

Reply:
I agree that for the case of 1 or 2 machines with an old delete timestamp, that could probably be a warning.  If the cluster-autoscaler or MHC are short-circuited because of excessive machines/nodes being unhealthy, that would be a critical.

If a control-plane machine is in anyway not healthy, that would be a critical.  We definitely need to iron out some of the details in some of our existing alerts and some of the new ones.

Reply:
> If the cluster-autoscaler or MHC are short-circuited because of excessive machines/nodes being unhealthy, that would be a critical.

So make a critical alert for "there is no schedulable space left"?  Maybe the cluster is still serving clients happily even without more space, and mitigation can wait another hour?

> If a control-plane machine is in anyway not healthy, that would be a critical.

This is one hiccup away from quorum loss, so agreed here.  But maybe etcd should be critical about quorum-vulnerable and CVO would be critical about "failing to schedule operator deployment" and the MHC warning would just inform recovery from those spout-ier critical issues?

Reply:
General statement to lili's comment (going even broader):

In a declarative system, whenever the declarative goal cannot be realized within time X, that is an alertable condition.  It may not be a great alert in all cases, but alerted >>> unalerted.  We can always improve the alert later.  However, the key criteria is "does not converge" is ALWAYS better to know about than not, because declarative systems depend on convergence to satisfy user intent.

I would like us to make sure we talk about this more broadly, but this is probably as important as "USE" to kubernetes alerting and metrics - "failure to converge" is the most important error in "E", and all teams should ask "do you alert when you do not converge within a reasonable amount of time".  I struggle to think of more than a handful of examples where failure to converge is non-alertable.